### PR TITLE
mixed-precision quantization milestone1: naive_intNwo + eval/benchmark framework

### PR DESCRIPTION
**Pull request by Hanxian97**
Summary:
This is a prototype for mixed-precision quantization. It consists of naive implementation of integer 2/3/5/6-bit quantization. Along with the int4wo and int8wo in torchao, it provides an evaluation framework leveraging lm_eval for mixed-precision quantization on Llama3

Test Plan:
To test the naive implementation of quantization APIs: python test/quantization/test_native_intNwo.py

### CODE REVIEW

File: torchao/quantization/prototype/mixed_precision/run_sensi_linear_type.sh [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    GPU="${GPUS[$j]}"
+
+    LOG_FILE="Sensi_skipsensi_${CONFIG1}_${CONFIG2}.txt"
+
+    CUDA_VISIBLE_DEVICES=$GPU python $PYTHON_SCRIPT --repo_id=checkpoints/meta-llama/Meta-Llama-3-8B --quantization=$CONFIG2 --linear_type=$CONFIG1 &>"$LOG_FILE" &
+  done
+
+  wait
+done
+
```
Comment (andrewor14) [inline torchao/quantization/prototype/mixed_precision/run_sensi_linear_type.sh:24]: Hi @Hanxian97, I feel we don't want to push these experiments scripts to torchao. Can you remove them from the PR? (OK to keep in your own separate branch for now)

File: torchao/quantization/prototype/mixed_precision/scripts/naive_intNwo.py [hunk 2]
```diff
@@ -0,0 +1,40 @@
+import torch
+
+from torchao.quantization.quant_primitives import (
+    MappingType,
+    ZeroPointDomain,
+)
+
+def intN_weight_only_asym(group_size=32, n=8):
```
Comment (andrewor14) [inline torchao/quantization/prototype/mixed_precision/scripts/naive_intNwo.py:8]: Can you add a short docstring to describe what this is doing? Maybe add an example to use this with the `quantize_` API? (same for `intN_weight_only_sym`)

Comment (andrewor14) [reply torchao/quantization/prototype/mixed_precision/scripts/naive_intNwo.py:8]: also, should we limit this to n = [2, 3, 4, 5, 6, 8] for now? (throw error otherwise)

File: torchao/quantization/prototype/mixed_precision/scripts/sensitivity_study.py [hunk 3]
```diff
@@ -0,0 +1,95 @@
+import torch
```
Comment (andrewor14) [inline torchao/quantization/prototype/mixed_precision/scripts/sensitivity_study.py:1]: might wanna call this file something else since you're about to do the real sensitivity analysis

File: torchao/quantization/prototype/mixed_precision/scripts/mp_quant_eval.py [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+torch._inductor.config.fx_graph_cache = True
+
+
+def run_evaluation(repo_id, tasks, limit, device, precision, quantization, compile, batch_size, max_length, sensi_bit, non_sensi_bit, quant_sym, group_size):
+
+    tokenizer = AutoTokenizer.from_pretrained(repo_id)
+    model = AutoModelForCausalLM.from_pretrained(repo_id).to(device="cpu", dtype=precision)
+
+    if quantization == "int8dq":
+        quantize_(model.to(device=device), int8_dynamic_activation_int4_weight())
```
Comment (andrewor14) [inline torchao/quantization/prototype/mixed_precision/scripts/mp_quant_eval.py:32]: this seems wrong? On main it's `int8_dynamic_activation_int8_weight`: https://github.com/pytorch/ao/blob/0e6c1224c6645426ef5dbb1817b2966a9c68ac84/scripts/hf_eval.py#L52.

Actually we can probably just delete this case?

**Review by andrewor14 — COMMENTED** [review]
Hi @Hanxian97, thanks for the PR. I think you can remove all the files except for `mp_quant_eval.py`, `naive_intNwo.py`, and `test_naive_intNwo.py`. Left a few minor comments other than that

File: torchao/quantization/prototype/mixed_precision/scripts/naive_intNwo.py [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+    def apply_intN_weight_only_quant_asym(weight):
+        # avoid circular dep
+        from torchao.dtypes import to_affine_quantized
+        mapping_type = MappingType.ASYMMETRIC
```
Comment (jerryzh168) [inline torchao/quantization/prototype/mixed_precision/scripts/naive_intNwo.py:12]: you can pass around this arg I think, to reduce code duplication

File: torchao/quantization/prototype/mixed_precision/scripts/naive_intNwo.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+        block_size = (1, group_size)
+        target_dtype = torch.int8
+        quant_min = 0
+        quant_max = 2**n-1
```
Comment (jerryzh168) [inline torchao/quantization/prototype/mixed_precision/scripts/naive_intNwo.py:16]: should target_dtype be `torch.uint8` for this?

File: torchao/quantization/prototype/mixed_precision/scripts/mp_quant_eval.py [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+        # filter for non-sensitive layers
+        def filter_fn_nonsen(child: torch.nn.Module, cur_fqn:str) -> bool:
+            return isinstance(child, nn.Linear) and not(any(skiplayer in cur_fqn for skiplayer in ['.0.', '.1.', '.2.', '.30.', '.31.']))
+        
+        if sensi_bit != 16:
+            # quantize the sensitive layers
+            if sensi_bit == 8:
+                quantize_(model.to(device=device), int8_weight_only(), filter_fn_sen)
+            elif  sensi_bit == 4: 
+                quantize_(model.to(device=device), int4_weight_only(group_size=group_size), filter_fn_sen)
```
Comment (jerryzh168) [inline torchao/quantization/prototype/mixed_precision/scripts/mp_quant_eval.py:65]: you could merge these logic into `intN_weight_only_asym` I think

File: torchao/quantization/prototype/mixed_precision/scripts/quant_model_size.py [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+    # Convert to gigabytes
+    total_size_gb = total_size_bytes / (1024 ** 3)
+    
+    return total_size_gb
+
+# Example usage
+num_elements = 250945664 #number of elements per Llama3 linear layer
+group_size = 32  # Example value, please adjust as needed
+bit_zeropoint = 2  # Example value, please adjust as needed
+bit_scale = 2  # Example value, please adjust as needed
```
Comment (jerryzh168) [inline torchao/quantization/prototype/mixed_precision/scripts/quant_model_size.py:23]: are these bytes?

File: torchao/quantization/prototype/mixed_precision/scripts/quant_model_size.py [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+    # Convert to bytes
+    total_size_bytes = total_size_bits / 8
+    
+    # Convert to gigabytes
+    total_size_gb = total_size_bytes / (1024 ** 3)
+    
+    return total_size_gb
+
+# Example usage
+num_elements = 250945664 #number of elements per Llama3 linear layer
```
Comment (jerryzh168) [inline torchao/quantization/prototype/mixed_precision/scripts/quant_model_size.py:20]: can this be calculated from the model instead of hardcoded? also I feel a better integration is just to fix and extend https://github.com/pytorch/ao/blob/5787e9e73f6493765052e3638bd0ae9103496bae/torchao/utils.py#L188

File: torchao/quantization/prototype/mixed_precision/scripts/sensitivity_study.py [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+    change_linear_weights_to_int4_woqtensors,
+    change_linear_weights_to_int8_dqtensors,
+    change_linear_weights_to_int8_woqtensors,
+    autoquant,
+)
+
+torch._inductor.config.force_fuse_int_mm_with_mul = True
+torch._inductor.config.fx_graph_cache = True
+
+def intN_weight_only(group_size=32, n=8):
```
Comment (jerryzh168) [inline torchao/quantization/prototype/mixed_precision/scripts/sensitivity_study.py:28]: I'd suggest to name this in more detail, since you have different dtypes and asymmetric/symmetric, in this case it's uintN_asymmetric_weight_only (or probably pass around asymmetric/symmetric as an argument)

File: torchao/quantization/prototype/mixed_precision/scripts/sensitivity_study.py [hunk 11]
```diff
@@ -0,0 +1,95 @@
...
+
+        mapping_type = MappingType.ASYMMETRIC
+        block_size = (1, group_size)
+        target_dtype = torch.uint8
+        quant_min = 0
+        quant_max = 2**n-1
+        eps = 1e-6
+        preserve_zero = False
+        zero_point_dtype = torch.bfloat16
+        zero_point_domain = ZeroPointDomain.FLOAT
```
Comment (jerryzh168) [inline torchao/quantization/prototype/mixed_precision/scripts/sensitivity_study.py:41]: why is this `FLOAT`?

File: torchao/quantization/prototype/mixed_precision/run_sensi_linear_type.sh (hunk 1 above)
Comment (Hanxian97) [reply torchao/quantization/prototype/mixed_precision/run_sensi_linear_type.sh]: Thanks for the comment. I have removed experiment scripts and only kept mp_quant_eval.py, naive_intNwo.py, and test_naive_intNwo.py.

File: torchao/quantization/prototype/mixed_precision/scripts/naive_intNwo.py (hunk 2 above)
Comment (Hanxian97) [reply torchao/quantization/prototype/mixed_precision/scripts/naive_intNwo.py]: Added the doctoring and assertion to limit [2,3,4,5,6,8] only

File: torchao/quantization/prototype/mixed_precision/scripts/sensitivity_study.py (hunk 3 above)
Comment (Hanxian97) [reply torchao/quantization/prototype/mixed_precision/scripts/sensitivity_study.py]: Removed this file for now and will commit the real sensitivity analysis  in milestone2

File: torchao/quantization/prototype/mixed_precision/scripts/mp_quant_eval.py (hunk 4 above)
Comment (Hanxian97) [reply torchao/quantization/prototype/mixed_precision/scripts/mp_quant_eval.py]: removed this for now since we will not use this

File: torchao/quantization/prototype/mixed_precision/scripts/test_naive_intNwo.py [hunk 12]
```diff
@@ -0,0 +1,27 @@
+import torch
```
[the remark this answers is not in the document]
Comment (Hanxian97) [reply torchao/quantization/prototype/mixed_precision/scripts/test_naive_intNwo.py]: Put the test_naive_intNwo.py under test/quantization now

File: torchao/quantization/prototype/mixed_precision/scripts/naive_intNwo.py (hunk 5 above)
Comment (Hanxian97) [reply torchao/quantization/prototype/mixed_precision/scripts/naive_intNwo.py:25]: Put symmetric or asymmetric into arg now

File: torchao/quantization/prototype/mixed_precision/scripts/naive_intNwo.py (hunk 6 above)
Comment (Hanxian97) [reply torchao/quantization/prototype/mixed_precision/scripts/naive_intNwo.py]: Fixed this.

File: torchao/quantization/prototype/mixed_precision/scripts/mp_quant_eval.py (hunk 7 above)
Comment (Hanxian97) [reply torchao/quantization/prototype/mixed_precision/scripts/mp_quant_eval.py]: merged them into intN_weight_only now

File: torchao/quantization/prototype/mixed_precision/scripts/sensitivity_study.py (hunk 10 above)
Comment (Hanxian97) [reply torchao/quantization/prototype/mixed_precision/scripts/sensitivity_study.py]: I passed asymmetric/symmetric as an argument and merged them into intN_weight_only

File: torchao/quantization/prototype/mixed_precision/scripts/naive_intNwo.py [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+        from torchao.dtypes import to_affine_quantized
+        mapping_type = MappingType.ASYMMETRIC
+        block_size = (1, group_size)
+        target_dtype = torch.uint8
+        quant_min = 0
+        quant_max = 2**n-1
+        eps = 1e-6
+        preserve_zero = False
+        zero_point_dtype = torch.bfloat16
+        zero_point_domain = ZeroPointDomain.FLOAT
```
Comment (andrewor14) [inline torchao/quantization/prototype/mixed_precision/scripts/naive_intNwo.py:33]: I think this should be ZeroPointDomain.INT. FLOAT is mainly for the optimized int4 tinygemm kernel right now

File: test/quantization/test_naive_intNwo.py [hunk 14]
```diff
@@ -0,0 +1,46 @@
+import torch
```
Comment (andrewor14) [inline test/quantization/test_naive_intNwo.py:1]: maybe call this `test_mixed_precision.py` to match your prototype folder and for your future test cases as well?

File: torchao/quantization/prototype/mixed_precision/scripts/naive_intNwo.py (hunk 5 above)
Comment (andrewor14) [reply torchao/quantization/prototype/mixed_precision/scripts/naive_intNwo.py:25]: I think Jerry means merging the two functions `apply_intN_weight_only_quant_asym` and `apply_intN_weight_only_quant_sym`, e.g. something like:

```
def apply_intN_weight_only_quant(weight: Tensor, symmetric: bool):
    # avoid circular dependency
    from torchao.dtypes import to_affine_quantized
    if symmetric:
        mapping_type = MappingType.SYMMETRIC
        quant_min = None
        quant_max = None
            ...  # other fields that are different between sym vs asym
    else:
        mapping_type = MappingType.ASYMMETRIC
        quant_min = 0
        quant_max = 2**n-1
    block_size = (1, group_size)
    eps = 1e-6
    ...  # other common fields
    return to_affine_quantized(...)
```

**Review by andrewor14 — APPROVED** [review]
Approving to unblock. Thanks!

File: torchao/quantization/prototype/mixed_precision/scripts/sensitivity_study.py (hunk 11 above)
Comment (Hanxian97) [reply torchao/quantization/prototype/mixed_precision/scripts/sensitivity_study.py]: In asymmetric quant, I found preserve_zero=True+ZeroPointDomain.INT has lower sqnr and worse accuracy than   preserve_zero=False+ZeroPointDomain.Float.

File: torchao/quantization/prototype/mixed_precision/scripts/quant_model_size.py (hunk 8 above)
Comment (Hanxian97) [reply torchao/quantization/prototype/mixed_precision/scripts/quant_model_size.py]: Yes these are in bytes. Have fixed this.

File: torchao/quantization/prototype/mixed_precision/scripts/quant_model_size.py (hunk 9 above)
Comment (Hanxian97) [reply torchao/quantization/prototype/mixed_precision/scripts/quant_model_size.py]: Yes this is a temporary solution for Llama3. Thanks for the suggestion! I will try to generalize it by extend the get_model_size_in_bytes.

File: torchao/quantization/prototype/mixed_precision/scripts/sensitivity_study.py (hunk 11 above)
Comment (Hanxian97) [reply torchao/quantization/prototype/mixed_precision/scripts/sensitivity_study.py]: Thanks for pointing it out. Just changed it to INT.

File: test/quantization/test_naive_intNwo.py (hunk 14 above)
Comment (Hanxian97) [reply test/quantization/test_naive_intNwo.py]: Renamed it.

File: test/quantization/test_mixed_precision.py [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+
+
+# test if the asymmetric and symmetric quantization API works with different bit widths
+for i in range(2, 9):
+    #test for asymmetric quantization
+    try:
+        test_weight_only_quant(i, False)
+        print(f"Test passed for {i}-bit using naive intNwo asymmetric quantization implementation")
+    except Exception as e:
+        print(f"Exception handled in test loop for {i}-bit asymmetric quantization. Details: {e}")
```
Comment (andrewor14) [inline test/quantization/test_mixed_precision.py:39]: might want to actually raise this exception too?

File: test/quantization/test_mixed_precision.py [hunk 16]
```diff
@@ -0,0 +1,46 @@
+import torch
+import torch.nn as nn
+
+import os
+import sys
+# append the path to the naive_intNwo.py file
+sys.path.append(os.path.join(os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))), "torchao/quantization/prototype/mixed_precision/scripts"))
```
Comment (andrewor14) [inline test/quantization/test_mixed_precision.py:7]: Does it work if you just add an empty `__init__.py` to `torchao/quantization/prototype/mixed_precision`? Then you won't need this line anymore?

Comment (Hanxian97) [reply test/quantization/test_mixed_precision.py]: Added __init__.py and removed the path append
